### PR TITLE
5.next: add cell events

### PR DESCRIPTION
**Pull request by LordSimal**
Refs: https://github.com/cakephp/cakephp/pull/17456

### CODE REVIEW

File: tests/TestCase/View/CellTest.php [hunk 1]
```diff
@@ -460,4 +460,24 @@ public function testACachedViewCellReRendersWhenGivenADifferentTemplate(): void
...
+    {
+        /** @var \TestApp\View\Cell\ArticlesCell $cell */
+        $cell = $this->View->cell('Articles::doEcho', ['msg1' => 'dummy', 'msg2' => ' message']);
+
+        $manager = $this->View->getEventManager();
+        $manager->on('Cell.beforeAction', function ($event, $eventCell, $action) use ($cell) {
+            $this->assertSame($eventCell, $cell);
+            $this->assertEquals('doEcho', $action);
+        });
+        $manager->on('Cell.beforeAction', function ($event, $eventCell, $action) use ($cell) {
```
Comment (ADmad) [inline tests/TestCase/View/CellTest.php]:
```suggestion
        $manager->on('Cell.afterAction', function ($event, $eventCell, $action) use ($cell) {
```

File: src/View/Cell.php [hunk 2]
```diff
@@ -170,8 +170,10 @@ public function render(?string $template = null): string
 
         $render = function () use ($template) {
             try {
+                $this->dispatchEvent('Cell.beforeAction', [$this, $this->action]);
```
Comment (ADmad) [inline src/View/Cell.php]: We could pass `$this->args` too to the events.

File: tests/TestCase/View/CellTest.php [hunk 3]
```diff
@@ -460,4 +460,24 @@ public function testACachedViewCellReRendersWhenGivenADifferentTemplate(): void
...
+            $this->assertSame($eventCell, $cell);
+            $this->assertEquals('doEcho', $action);
+        });
+        $this->assertStringContainsString('dummy message', $cell->render());
```
Comment (ADmad) [inline tests/TestCase/View/CellTest.php]: This assertion is unrelated to the events' execution, can be changed to just `$cell->render()`.

Also for the asserts you have added there won't be any failure reported even if the events are not triggered. To check that the callbacks are run set a var outside the callback, pass it by ref to the callback and modify inside it and then assert it's value outside the callback.

File: tests/TestCase/View/CellTest.php (hunk 1 above)
Comment (LordSimal) [reply tests/TestCase/View/CellTest.php]: whoops...

**Pull request merged**